File: src/Midi2Image.java
import java.awt.Color;
import java.awt.Graphics2D;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;

import javax.imageio.ImageIO;
import javax.sound.midi.InvalidMidiDataException;
import javax.sound.midi.MetaMessage;
import javax.sound.midi.MidiEvent;
import javax.sound.midi.MidiMessage;
import javax.sound.midi.MidiSystem;
import javax.sound.midi.Sequence;
import javax.sound.midi.ShortMessage;
import javax.sound.midi.Track;

public class Midi2Image {
           
	private int rollDpi = 300;                                                                 
	private int rollTempo = 95;                                                                
	private int rollWidth = (int) (11.25 * this.rollDpi);  // @300dpi                          
	private int rollMargin = 50; // @300dpi                                                    
	private int holeWidth = 18; // @300dpi                                                     
	private int hole0Center = 36; // @300dpi                                                   
	private int hole99Center = 3340; // @300dpi                                                
	private int rollColor = 120; // gray scale              
	private int reducePX = 10;                                                       
	int[][] controlChange2NoteNo = {{64, 18}, {67, 113}};  // Control Change No to Hole No. Sustain/Soft pedals.              
                                                                                               
	private BufferedImage out_img = null;
	private Graphics2D g2d = null;
	       
	Midi2Image(){                    
	}
	
	protected int getHoleX(int note_no) {
		// Return hole x position in pixel by note number 
		return rollMargin + hole0Center + ((note_no - 15) * (hole99Center - hole0Center) / 99) - (holeWidth / 2);
	}

	protected int getTick2Px(long tickLen, int tempo, double bpm, int ppq) {
		// Return hole length in pixel by MIDI tick length
		return (int) ((tickLen * rollDpi * tempo * 1.2) / (bpm * ppq));
	}

	protected void drawHole(int note, int tempo, double bpm, int ppq, long onTick, long offTick) {
		int hole_h = getTick2Px(offTick - onTick, tempo, bpm, ppq) - reducePX;
		hole_h = hole_h < holeWidth ? holeWidth : hole_h; 
		int hole_x = getHoleX(note);
		int hole_y = out_img.getHeight() - getTick2Px(onTick, tempo, bpm, ppq) - hole_h;
		int hole_y_margin = (int) holeWidth / 3;
		int chain_perforation_th_len = 85;
		g2d.setColor(Color.WHITE);
		
		// Chain Perforation
		int y = hole_y;
		for (; y < hole_y + hole_h - chain_perforation_th_len; y += hole_y_margin + holeWidth) {
			g2d.fillOval(hole_x, y, holeWidth, holeWidth);
		}
		// Normal Perforation
		g2d.fillRoundRect(hole_x, y, holeWidth, hole_h - (y - hole_y), holeWidth, holeWidth);
		// somewhow, fillOval() result pops out 1px on left, so fill it.
		g2d.setColor(new Color(rollColor, rollColor, rollColor));
		g2d.drawLine(hole_x, hole_y, hole_x, hole_y + hole_h);
	}
	
	public void convert(String midiPath, String outDir) {
		try {
			Sequence sequence = MidiSystem.getSequence(new File(midiPath));
			int ppq = sequence.getResolution(); // PPQ (ticks per quarter note)
			long totalTickLen = sequence.getTickLength();

			double bpm = 80;
			int tempo = rollTempo;
			
			for (Track track : sequence.getTracks()) {
				long[] noteOnTicks = new long[128];
				for (int i = 0; i < track.size(); i++) {
					MidiEvent evt = track.get(i);
					MidiMessage msg = evt.getMessage();

					if (out_img == null && msg instanceof MetaMessage meta && meta.getType() == 0x51) {
						// Get BPM/Tempo
						byte[] data = meta.getData();
						int tempoMicro = ((data[0] & 0xFF) << 16) | ((data[1] & 0xFF) << 8) | (data[2] & 0xFF);
						bpm = 60_000_000.0 / tempoMicro;
						if (tempo == 0)
							tempo = (int) Math.round(bpm);
						System.out.printf("Tempo event at tick %d: %.2f BPM %d tempo\n", evt.getTick(), bpm, tempo);
						// Create Image Background
						int width = rollWidth + 2 * rollMargin;
						int img_h = getTick2Px(totalTickLen, tempo, bpm, ppq);
						out_img = new BufferedImage(width, img_h, BufferedImage.TYPE_BYTE_GRAY);
						g2d = out_img.createGraphics();
						// Draw Roll Edge
						g2d.setColor(Color.WHITE);
						g2d.fillRect(0, 0, rollMargin, img_h);
						g2d.fillRect(rollWidth + rollMargin, 0, rollMargin, img_h);
						// Draw Roll
						g2d.setColor(new Color(rollColor, rollColor, rollColor));
						g2d.fillRect(rollMargin, 0, rollWidth, img_h);
						g2d.setColor(Color.WHITE);
					}

					if (msg instanceof ShortMessage sm) {
						int cmd = sm.getCommand();
						int note = sm.getData1();
						int velocity = sm.getData2();
						long tick = evt.getTick();

						// Pedal Control Change to Note No
						for (int[] pair : controlChange2NoteNo) {
						    int cc_number = pair[0];
						    int note_numer = pair[1];
							if (cmd == ShortMessage.CONTROL_CHANGE && note == cc_number && velocity > 0) {
								noteOnTicks[note_numer] = tick;
							} else if (cmd == ShortMessage.CONTROL_CHANGE && note == cc_number && velocity == 0) {
								long onTick = noteOnTicks[note_numer];
								drawHole(note_numer, tempo, bpm, ppq, onTick, tick);
							}
						}

						// Note On/Off
						if (cmd == ShortMessage.NOTE_ON && velocity > 0) {
							noteOnTicks[note] = tick;
						} else if (cmd == ShortMessage.NOTE_OFF || cmd == ShortMessage.NOTE_ON && velocity == 0) {
							long onTick = noteOnTicks[note];
							drawHole(note, tempo, bpm, ppq, onTick, tick);
						}
					}
				}
			}	
			// save image
			String save_name = outDir + (new File(midiPath)).getName().replace(".mid", String.format(" tempo%d.png", tempo));
			ImageIO.write(out_img, "png", new File(save_name));
		} catch (InvalidMidiDataException | IOException e) {
			e.printStackTrace();
		}	
	}

	public static void main(String[] args) {

		File dir = new File("C:\\Users\\sasaki\\Downloads\\Ampico_All_erolls\\test\\");
		for (File file : dir.listFiles()) {
		    if (file.isDirectory()) continue;
		    
			var obj = new Midi2Image();
			obj.convert(file.getPath(), "output/classic/");
		}	    		
	}
}

File: src/Midi2Image.py
import os
import time

from mido import MidiFile
from PIL import Image, ImageDraw


class Midi2Image:
    def __init__(self):
        self.roll_dpi = 300
        self.roll_tempo = 95
        self.roll_width = int(11.25 * self.roll_dpi)
        self.roll_margin = 50
        self.hole_width = 18
        self.hole_y_margin = self.hole_width // 3
        self.hole_0_center = 36
        self.hole_99_center = 3340
        self.roll_color= 120  # grayscale
        self.reduce_px = 10
        self.chain_perforation_th_len = 85
        self.control_change_map = [
            {"controlChangeNo": 64, "midiNoteNo": 18},  # ControlChangeMap(64, 18)
			{"controlChangeNo": 67, "midiNoteNo": 113},	
		]
        self.hole_x_list = [self._get_hole_x(i) for i in range(128)]
        self.out_img = None
        self.draw = None

    def _get_hole_x(self, note_no):
        return int(self.roll_margin + self.hole_0_center + ((note_no) * (self.hole_99_center - self.hole_0_center) / 99) - (self.hole_width / 2))

    def get_tick_to_px(self, tick_len, tempo, bpm, ppq):
        return int((tick_len * self.roll_dpi * tempo * 1.2) / (bpm * ppq))

    def draw_hole(self, note_no, tempo, bpm, ppq, on_tick, off_tick):
        hole_h = self.get_tick_to_px(off_tick - on_tick, tempo, bpm, ppq) - self.reduce_px
        hole_h = max(hole_h, self.hole_width)
        hole_x = self.hole_x_list[note_no - 15]
        hole_y = self.out_img.height - self.get_tick_to_px(on_tick, tempo, bpm, ppq) - hole_h

		# Chain Perforation
        y = hole_y
        while y < hole_y + hole_h - self.chain_perforation_th_len:
            self.draw.ellipse([hole_x, y, hole_x + self.hole_width, y + self.hole_width], fill=255)
            y += self.hole_y_margin + self.hole_width

        # Normal perforation
        self.draw.rounded_rectangle([hole_x, y, hole_x + self.hole_width, hole_y + hole_h], radius=self.hole_width // 2, fill=255)

    def convert(self, midi_path, out_dir):
        mid = MidiFile(midi_path)
        ppq = mid.ticks_per_beat
        bpm = 80.0
        tempo = self.roll_tempo

        note_on_ticks = [0] * 128
        initialized = False

        for track in mid.tracks:
            abs_tick = 0
            for msg in track:
                abs_tick += msg.time

                if msg.type == "set_tempo" and not initialized:
                    tempo_micro = msg.tempo
                    bpm = 60_000_000.0 / tempo_micro
                    if tempo == 0:
                        tempo = int(round(bpm))
                    print(f"Tempo event at tick {abs_tick}: {bpm:.2f} BPM")

                    # 画像の初期化
                    width = self.roll_width + 2 * self.roll_margin
                    total_ticks = sum([t.time for t in track])
                    img_h = self.get_tick_to_px(total_ticks, tempo, bpm, ppq)
                    self.out_img = Image.new("L", (width, img_h), color=self.roll_color)
                    self.draw = ImageDraw.Draw(self.out_img)
                    self.draw.rectangle([0, 0, self.roll_margin, img_h], fill=255)
                    self.draw.rectangle([self.roll_margin + self.roll_width, 0, width, img_h], fill=255)
                    initialized = True

                if msg.type in ("note_on", "note_off", "control_change"):
                    note = msg.note if hasattr(msg, "note") else msg.control
                    velocity = msg.velocity if hasattr(msg, "velocity") else msg.value

                    for map in self.control_change_map:
                        if msg.type == "control_change" and note == map.controlChangeNo:
                            if velocity > 0:
                                note_on_ticks[map.midiNoteNo] = abs_tick
                            else:
                                self.draw_hole(map.midiNoteNo, tempo, bpm, ppq, note_on_ticks[map.midiNoteNo], abs_tick)

                    if msg.type == "note_on" and msg.velocity > 0:
                        note_on_ticks[note] = abs_tick
                    elif msg.type == "note_off" or  (msg.type == "note_on" and msg.velocity == 0):
                        self.draw_hole(note, tempo, bpm, ppq, note_on_ticks[note], abs_tick)

        if self.out_img:
            save_name = os.path.join(out_dir, os.path.basename(midi_path).replace(".mid", f" tempo{tempo}.png"))
            os.makedirs(os.path.dirname(save_name), exist_ok=True)
            self.out_img.save(save_name)
            print(f"Saved image to {save_name}")

if __name__ == "__main__":
    input_dir = "C:\\Users\\sasaki\\Downloads\\Ampico_All_erolls\\test\\"
    output_dir = "output/classic/"
    for filename in os.listdir(input_dir):
        if filename.lower().endswith(".mid"):
            path = os.path.join(input_dir, filename)
            t1 = time.time()
            m2i = Midi2Image()
            m2i.convert(path, output_dir)
            print(time.time() - t1, "sec")